[WIP] [TACHYON-221] Modularize out Under File System implementations (Attempt #2)

rvesse opened the pull request:
This is a first pass pull request to gather comments on the proposed changes that modularise out the Under File System implementations as far as I have been meaningfully able to.

This has been updated from pull request #642 to rebase against the latest master which has the hierarchy store improvements merged in

### Overview

The current code base has three implementations:

* Local
* HDFS
* Gluster (which is conflated into HDFS implementation at the moment)

This pull request leaves the local implementation in the core module because we need something for basic integration testing though it does move it into a dedicated `tachyon.underfs` package.  It then pulls the other two implementations into their own modules, actually 3 new Maven modules are added:

* `underfs`
    * `underfs-hdfs`
    * `underfs-glusterfs`

The extra parent module is used to logically group the modules within the source tree and provides a useful future home for any generic under file system code that may be abstracted out of core in the future.

`UnderFileSystem` instances are now created indirectly using a new `UnderFileSystemFactory`, this provides a means to determine if a factory supports a given path and if so to create an instance.  Factories are registered and managed by the `UnderFileSystemRegistry` which primarily auto-discovers available implementations using the `ServiceLoader` mechanism to discover implementations declared in the `META-INF\services\tachyon.underfs.UnderFileSystemFactory` file.  The registry can also be used to manually register factories in the case where the automatic discovery proves insufficient.  The existing `UnderFileSystem.get()` methods remain but these now use the `UnderFileSystemRegistry` to determine which implementation to use rather than hard coding that decision which allows much more flexibility e.g. substituting in alternative `UnderFileSystem` implementations for HDFS.

### Testing

One advantage of this modularisation is that a single build can now potentially test all the under file system implementations because each is located in a different module.  The POMs of the new modules use the surefire plugin to pull in the tests from the core module and re-run them with their different implementations.  As a result the `underfs-hdfs` module will now always run HDFS backed tests (unless tests are being skipped) since we are able to start a miniature HDFS cluster in-memory so it is no longer necessary to have a separate profile for these tests.

However in the case of `underfs-gluster` the `glusterfsTest` profile must still be explicitly activated as we don't expect users to have Gluster FS available on their system nor do we have the ability to emulate it as we can with HDFS.

Bottom line is that the build with tests will now fully test both local and HDFS under file system implementations.

### Maven

A side effect of the refactoring was that certain plugin configuration files were pulled up under a `build` directory in the top level directory (Checkstyle, FindBugs, License Header) and the plugin configurations appropriately replicated to the new modules.  This bit could potentially be cleaned up to allow the plugins to be defined at a higher level to avoid replicating some configuration though while the number of modules remains small this may not be such an issue.

One new addition to the POMs is that license headers are now enforced as far as I thought was meaningful (for example Thrift generated code is excluded) using the `license-maven-plugin` which is useful to ensure consistency in headers and avoid missing headers.  As a result of this there are some minor formatting changes to some headers to ensure that they are detected correctly.

### TODOs

There were a few places where I couldn't simplify the existing code as much as I would have liked because of the need to support pre 2.x Hadoop versions (e.g. the `isHadoopUnderFS()` method could otherwise be greatly simplified/eliminated).  There are also a couple of unit test specific pieces of codes that it would be nice to generalise but it was unclear exactly how this could be done.

`TODO` comments are added in to indicate these options

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/2672/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/2673/<h2>Build result: ABORTED</span></h2>[...truncated 2112 lines...]Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 95.944 sec - in tachyon.client.FileInStreamTestRunning tachyon.client.LocalBlockInStreamTestTests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 14.336 sec - in tachyon.client.LocalBlockInStreamTestRunning tachyon.client.RemoteBlockInStreamTestTests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 62.168 sec - in tachyon.client.RemoteBlockInStreamTestRunning tachyon.client.TachyonFSTestTests run: 30, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 12.63 sec - in tachyon.client.TachyonFSTestRunning tachyon.client.BlockInStreamTestTests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 50.016 sec - in tachyon.client.BlockInStreamTestRunning tachyon.client.FileOutStreamTestTests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 19.663 sec - in tachyon.client.FileOutStreamTestRunning tachyon.client.table.RawColumnTestTests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 7.909 sec - in tachyon.client.table.RawColumnTestRunning tachyon.client.table.RawTableTestBuild timed out (after 60 minutes). Marking the build as aborted.Results :Tests run: 336, Failures: 0, Errors: 0, Skipped: 1[ERROR] There was a timeout or other error in the forkBuild was abortedchannel stopped[WARNING] Failed to notify spy hudson.maven.Maven3Builder$JenkinsEventSpy: hudson.remoting.Channel$OrderlyShutdown
Test FAILed.

rvesse commented:
Rebase seems to have broken my branch more than I realised, closing while I figure out the issues though please feel free to look over the code and comment on it should anyone have time to review this

rvesse commented:
Reopening, build issues locally were down to an incomplete clean on my system.  Build issues on Jenkins appears to be due to timeout.

As with this PR the full set of tests are now run over both a local and HDFS under file system the build now takes longer than this.   Locally the build takes just over an hour on my system and Jenkins appears to be configured with a 60 minute timeout.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/2674/
Test PASSed.

hsaputra commented:
Hi @rvesse, is the changes to checkstyle files needed for this patch?

Prefer to have isolated patch to help review.

haoyuan commented:
@hsaputra 
+1 on a isolated patch.

rvesse commented:
@hsaputra @haoyuan Sorry I had not meant to commit the Eclipse generated `.checkstyle` files and I have removed this from the PR now

The only "change" to the actual check style files is to move them to a top level location to make it easier to re-use them across all modules

rvesse commented:
I should not even once you guys have a chance to review this please do not merge this yet even if it looks good so far (though I suspect you'll have some feedback that needs addressing).

There are a couple of tweaks I still need to make:

* Support considering multiple candidate `UnderFileSystemFactory` instances - this supports scenarios where you might have multiple factories support the same style of path but some can only create clients in specific circumstances
* Make sure the uber JAR used to run Tachyon processes pulls in the underfs modules

dcapwell commented:
Been away for a while, so still catching up to all changes that have come in.  I'll look at this more later today.

Can you split the license work into its own PR?

rvesse commented:
@dcapwell Which bits would you like splitting out?  Just the addition of the license plugin or all the fix ups to license headers?

dcapwell commented:
Unless I'm reading the patch wrong (too large for phone so skimmed real
quick), the work is to refactor ufs, but then there is logic for header
checks; world be good for the header checks to be split out.

Might have misread the PR, doing it on my phone so context gets harder
On Feb 6, 2015 9:23 AM, "Rob Vesse" <notifications@github.com> wrote:

> @dcapwell <https://github.com/dcapwell> Which bits would you like
> splitting out? Just the addition of the license plugin or all the fix ups
> to license headers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/647#issuecomment-73276959>.
>

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tachyon-Pull-Request-Builder/2676/
Test PASSed.

dcapwell commented on core/src/main/java/tachyon/client/TachyonFS.java:
```diff
@@ -96,15 +101,18 @@ public static synchronized TachyonFS get(final TachyonURI tachyonURI) throws IOE
    */
-  public static synchronized TachyonFS get(String masterHost, int masterPort, boolean zookeeperMode)
-      throws IOException {
+  public static synchronized TachyonFS
```
formatting seems wrong, can you correct this.

dcapwell commented on core/src/main/java/tachyon/master/TachyonMaster.java:
```diff
@@ -186,20 +185,13 @@ boolean isZookeeperMode() {
-      ((UnderFileSystemHdfs) ufs).login(mConf.KEYTAB_KEY, mConf.KEYTAB, mConf.PRINCIPAL_KEY,
-          mConf.PRINCIPAL, NetworkUtils.getFqdnHost(mMasterAddress));
-    }
+    ufs.connectFromMaster(NetworkUtils.getFqdnHost(mMasterAddress));
```
Since this is HDFS only, this logic should only be in the HDFS class.  Can you move that there?  I thought that that was the case before, but guess that didn't happen...

dcapwell replied:
aka, not create this function for now (even login)

rvesse replied:
I didn't want to preclude that other future implementations might need to take some actions to establish a connection.  The actual login logic has moved to these methods in the HDFS implementation.

The logic can't be confined to the HDFS implementation constructor (i.e. rendering these methods unnecessary) because logging into HDFS requires knowing the host we want to present ourselves as.  While we could try and auto-detect the host name of the system there is no guarantee the host name we'd find is the externally visible host name which is what we actually need.

dcapwell commented on core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    // Discover and register the available factories
+    ServiceLoader<UnderFileSystemFactory> discoverableFactories =
```
ServiceLoader implements Iterable, so woud this not be best to do with for expression?

```
for (UnderFileSystemFactory factory : discoverableFactories)
```
```

rvesse replied:
Good point

dcapwell commented on core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java:
```diff
@@ -0,0 +1,216 @@
+public class UnderFileSystemRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final List<UnderFileSystemFactory> FACTORIES =
```
Might be better to use a map with key of schema.  That way find is just a check in the map rather than iterate over.

rvesse replied:
The motivation here was not to preclude there being multiple implementations supporting the same scheme or that rely on additional information encoded in the URL or that they expect to have provided via the `conf` object passed to `create()`

dcapwell commented on core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java:
```diff
@@ -0,0 +1,216 @@
+    init();
+  }
+
+  static synchronized void init() {
```
Can you make this private?

rvesse replied:
Yep

dcapwell commented on core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java:
```diff
@@ -0,0 +1,216 @@
+   */
+  public static synchronized void reset() {
+    // Can't reset if we're not initialized
+    if (!sInit) {
```
When would this be the case?  Currently all methods put a lock on the class object, so only one function can play around with mutating `FACTORIES`.

If there is a case where this would happen, wouldn't it be better to call init rather than returning?

rvesse replied:
Again good point

dcapwell commented on core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java:
```diff
@@ -0,0 +1,216 @@
+
+    LOG.debug("Registered Under File System Factory implementation {} - {}", factory.getClass(),
+        factory.toString());
+    FACTORIES.add(0, factory);
```
Never used `CopyOnWriteArrayList` before.  http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CopyOnWriteArrayList.html#add(int,%20E)  Says that it will insert at head (in this case) and move everything to the right.  For this data struct, is that faster than calling `add(factory)`?

rvesse replied:
The intention of using `CopyOnWriteArrayList` iwas to avoid having to explicitly lock on `FACTORIES` and not to worry if someone is changing its definition while other methods are being called

Implementation wise I intentionally insert at the head because the point of this method is to allow users to manually override the automatically discovered implementations where necessary.

dcapwell replied:
Thanks, can you leave a comment in the code then?
On Feb 6, 2015 3:58 PM, "Rob Vesse" <notifications@github.com> wrote:

> In core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java
> <https://github.com/amplab/tachyon/pull/647#discussion_r24281776>:
>
> > +   * it should be unnecessary to call this, however since ServiceLoader discovery order may be
> > +   * susceptible to class loader behavioural differences there may be rare cases when you need to
> > +   * manually register the desired factory.
> > +   * <p>
> > +   *
> > +   * @param factory
> > +   *          Factory to add
> > +   */
> > +  public static void add(UnderFileSystemFactory factory) {
> > +    if (factory == null) {
> > +      return;
> > +    }
> > +
> > +    LOG.debug("Registered Under File System Factory implementation {} - {}", factory.getClass(),
> > +        factory.toString());
> > +    FACTORIES.add(0, factory);
>
> The intention of using CopyOnWriteArrayList iwas to avoid having to
> explicitly lock on FACTORIES and not to worry if someone is changing its
> definition while other methods are being called
>
> Implementation wise I intentionally insert at the head because the point
> of this method is to allow users to manually override the automatically
> discovered implementations where necessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/647/files#r24281776>.
>

dcapwell commented on core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java:
```diff
@@ -0,0 +1,216 @@
+    // Try to obtain the appropriate factory
+    UnderFileSystemFactory factory = find(path);
+    if (factory == null) {
+      throw new IllegalArgumentException(String.format(
```
In this case, woudln't string concat be faster and easier to read?

rvesse replied:
Sure, this is just my stylistic preference but I'm happy to go with whatever is preferred

dcapwell commented on core/src/main/java/tachyon/underfs/local/SingleMachineLocalUnderFileSystem.java:
```diff
@@ -33,11 +35,7 @@
-  public static UnderFileSystem getClient() {
-    return new UnderFileSystemSingleLocal();
-  }
+public class SingleMachineLocalUnderFileSystem extends UnderFileSystem {
```
Why `SingleMachine`?  If you point this to a linux (osx) mounted file system that is distributed, its no longer single machine.  `LocalUnderFileSystem` seems good to me.

rvesse replied:
Just because it was originally had `Single` in its name, however I agree that `LocalUnderFileSystem` is a more sensible name

dcapwell commented on core/src/main/java/tachyon/worker/WorkerStorage.java:
```diff
@@ -245,8 +245,10 @@ public void run() {
           mMasterClient.addCheckpoint(mWorkerId, fileId, fileSizeByte, dstPath);
-          long shouldTakeMs = (long) (1000.0 * fileSizeByte / Constants.MB
-              / WorkerConf.get().WORKER_PER_THREAD_CHECKPOINT_CAP_MB_SEC);
+          long shouldTakeMs =
```
Don't think this follows the style, but I find this change more readable.

rvesse replied:
Maven Checkstyle failed the build for me though maybe only because applying the Tachyon formatter put that line onto a single line which may be the formatter not being quite right

dcapwell commented:
Ok, I got about 70% through and I would say 90% of the patch doesn't have anything to do with UFS.  Could you revert the format changes so its easier to review the UFS changes?  


I like the idea of using `ServiceLocator` so its easier to split.  One thing that I didn't get to see is how you merge the files together.  Uber jars and `ServiceLocator` don't really like each other...

rvesse commented:
Most of the formatting changes are simply me applying Eclipse reformat with the Tachyon formatter so any formatter changes were things that weren't previously adhering to the formatter

As for `ServiceLoader` and uber-jar all the current shade invocations explicitly use the `ServicesResourceTransformer` which handles merging `META-INF\services` files together nicely

I will close this PR while I address the various comments and extract out some of the irrelevant stuff where possible

dcapwell commented:
Thanks for pointing out that shade has this already!

There have been so many issues around formatting with this project.  Many
new people setup eclipse and idea to do it but then things go wrong.

Hy, formatting should be relocked at since this is happening very often.
Can you file a JIRA for this? On phone so don't have access
On Feb 6, 2015 4:02 PM, "Rob Vesse" <notifications@github.com> wrote:

> Most of the formatting changes are simply me applying Eclipse reformat
> with the Tachyon formatter so any formatter changes were things that
> weren't previously adhering to the formatter
>
> As for ServiceLoader and uber-jar all the current shade invocations
> explicitly use the ServicesResourceTransformer which handles merging
> META-INF\services files together nicely
>
> I will close this PR while I address the various comments and extract out
> some of the irrelevant stuff where possible
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/647#issuecomment-73335963>.
>

rvesse commented:
License header changes and enforcement extracted as #650

dcapwell commented:
Thanks
On Feb 6, 2015 5:05 PM, "Rob Vesse" <notifications@github.com> wrote:

> License header changes and enforcement extracted as #650
> <https://github.com/amplab/tachyon/pull/650>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/647#issuecomment-73341358>.
>

rvesse commented:
Also now opened #653 which extracts out the check style and find bugs changes.

Once that gets reviewed and merged I'll rebase this PR onto master squashing out the irrelevant changes as much as possible which should leave this as only code related (bar the earlier comments about formatting issues)

rvesse commented:
@dcapwell There already appears to be a JIRA about the formatter issue, see [TACHYON-243](https://tachyon.atlassian.net/browse/TACHYON-243) and there is an open PR #554 that addresses some issues around this

dcapwell commented:
Thanks
On Feb 9, 2015 2:45 AM, "Rob Vesse" <notifications@github.com> wrote:

> @dcapwell <https://github.com/dcapwell> There already appears to be a
> JIRA about the formatter issue, see TACHYON-243
> <https://tachyon.atlassian.net/browse/TACHYON-243> and there is an open
> PR #554 <https://github.com/amplab/tachyon/pull/554> that addresses some
> issues around this
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/647#issuecomment-73488972>.
>
